sockstat: Add automatic column sizing and remove -w option

daminrido139 opened the pull request:
Refactor sockstat to dynamically size table columns based on content. This change eliminates the need for the -w option, which is now ignored for backwards compatibility. Column widths are calculated per flag, improving accuracy and overall performance.

Sponsored by: Google, LLC (GSoC 2025)
MFC after: 2 weeks

daminrido139 commented:
Thanks, @asomers,

You're absolutely right, `"stream (not connected)"` should be fully inside the PROTO column. I initially misunderstood and treated `(not connected)` as part of the LOCAL ADDRESS, but I’ll fix that and ensure the width calculation includes the full string.

Regarding the `"??"` change - yes, that was intentional. I wanted to make the output more uniform by using `"??"` for all unknown fields instead of mixing `""` and `"?"`. Sorry for not mentioning that in the commit message. happy to amend the commit if you prefer.

asomers commented:
> Thanks, @asomers,
> 
> You're absolutely right, `"stream (not connected)"` should be fully inside the PROTO column. I initially misunderstood and treated `(not connected)` as part of the LOCAL ADDRESS, but I’ll fix that and ensure the width calculation includes the full string.
> 
> Regarding the `"??"` change - yes, that was intentional. I wanted to make the output more uniform by using `"??"` for all unknown fields instead of mixing `""` and `"?"`. Sorry for not mentioning that in the commit message. happy to amend the commit if you prefer.

Using "??" uniformly is ok.  The only risk is that it upsets some scripts that people use to parse sockstat's output with awk.  But those scripts are fragile anyway.  Using libxo would be much more reliable!

daminrido139 commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1108,234 +1094,428 @@ sctp_path_state(int state)
+				printf(" %-*s", cw->local_addr, "(not connected)");
+			strcpy(buf,"??");
+			if (laddr->address.ss_len > 0)
+				formataddr(&laddr->address, buf, sizeof(buf));
```
fixed!

daminrido139 commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1344,33 +1524,38 @@ display(void)
 	struct sock *s;
-	int n, pos;
+	int n;
+    struct col_widths cw;
```
fixed!

asomers reviewed (commented):
Pro tips:
* Once a reviewer first looks at a PR, never force-push it.  If you must amend
  it, leave your changes in a separate commit.  That makes it easy for the
  reviewer to see what's changed since his first review.
* Never merge the main branch into your PR branch.  That can confuse the
  history and doesn't really solve any problems.  Instead, better to leave the
  PR out-of-date unless it conflicts with the main branch.  If you must resolve
  conflicts, it's better to rebase than to merge.

I'll come back tomorrow for a lengthier review.  I'll have to re-review the entire thing.

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		if (opt_i) {
+			if (s->proto == IPPROTO_TCP ||
+				s->proto == IPPROTO_UDP) {
+				printf(" %*" PRIu64, cw->inp_gencnt ,s->inp_gencnt);
```
Style: line longer than 80 chars

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+			printf(" %-*s", cw->splice_address, buf);
+		}
+		if (opt_i) {
+			if (s->proto == IPPROTO_TCP ||
```
These two lines would be longer than 80 chars if you joined them.  However, it's acceptable to join them and then move the `{` to the next line.  That's more readable, IMHO.

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+			if (s->proto == IPPROTO_TCP ||
+				s->proto == IPPROTO_UDP) {
+				printf(" %*" PRIu64, cw->inp_gencnt ,s->inp_gencnt);
+			}
+			else
```
style: cuddle the `else`
```suggestion
			} else
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+				if (sp != NULL)
+					formataddr(&sp->laddr->address,
+								buf, bufsize);
+			}
+			else
```
style: cuddle the else
```suggestion
			} else
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+					s->state != TCPS_CLOSED &&
+					s->state != TCPS_LISTEN))) {
+				printf(" %*u", cw->encaps, ntohs(faddr->encaps_port));
+			}
+			else
```
```suggestion
			} else
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+				s->state != SCTP_LISTEN) {
+				printf(" %-*s", cw->path_state,
+					sctp_path_state(faddr->state));
+			}
+			else
```
```suggestion
			} else
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
 						break;
 					}
 				}
-				offset += 13;
+				else
```
style: cuddle the `else`

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-				}
-				offset += TCP_FUNCTION_NAME_LEN_MAX + 1;
+				if (s->proto == IPPROTO_TCP)
+					printf(" %-*.*s", TCP_FUNCTION_NAME_LEN_MAX,
+						TCP_FUNCTION_NAME_LEN_MAX, s->stack);
+				else
+					printf(" %-*s", cw->stack, "??");
```
For both brevity and consistency
```suggestion
					printf(" %-*.*s", cw->stack, cw->stack, s->stack);
				else
					printf(" %-*s", cw->stack, "??");
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-				}
-				offset += TCP_CA_NAME_MAX + 1;
+				if (s->proto == IPPROTO_TCP)
+					printf(" %-*.*s", TCP_CA_NAME_MAX,
+						TCP_CA_NAME_MAX, s->cc);
```
```suggestion
					printf(" %-*.*s", cw->cc, cw->cc, s->cc);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		first = false;
 	}
-	xprintf("\n");
+	printf("\n");
```
Pro tip: using `printf` repeatedly is inefficient, because every call results in a `write(2)` syscall.  Using `fprintf` is much more efficient if you plan to call it over and over, because it buffers the output in the C library, and only calls `write(2)` when the buffer is full or you intentionally flush it.  HOWEVER, don't make that change now, because in a future commit you'll replace all of these with `xo_emit` instead.

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+	/* Remote peer we connect(2) to, if any. */
+	if (faddr->conn != 0) {
+		struct sock *p;
+		strlcpy(buf, "-> ", bufsize);
+		pos += 3;
```
Better to rely on strlcpy returning the right value than to hard-code it.
```suggestion
		pos += strlcpy(buf, "-> ", bufsize);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+				&(struct file){ .xf_data =
+				p->socket });
+			if (f != NULL) {
+				pos += snprintf(buf == NULL ? NULL : buf+pos,
+					buf == NULL ? 0 : bufsize-pos,
+					"[%lu %d]", (u_long)f->xf_pid, f->xf_fd);
```
Instead of trying to keep track of `pos`, you could print lines like this into a stack buffer, then use `strlcat` to copy that to `buf`.  Which technique you prefer is your choice.

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+	if (laddr != NULL && faddr != NULL && s->family == AF_UNIX &&
+		laddr->address.ss_len == 0 && faddr->conn == 0)
+		len += 16;
+	if (len > cw->proto)
+		cw->proto = len;
```
```suggestion
	cw->proto = MAX(cw->proto, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		len += 1;
+	if (laddr != NULL && faddr != NULL && s->family == AF_UNIX &&
+		laddr->address.ss_len == 0 && faddr->conn == 0)
+		len += 16;
```
Where does the magic number 16 come from?

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		} else {
+			if (laddr != NULL) {
+				len = formataddr(&laddr->address, NULL, 0);
+				if (len > cw->local_addr)
+					cw->local_addr = len;
```
```suggestion
				cw->local_addr = MAX(cw->local_addr, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-				}
+			if (faddr != NULL) {
+				len = formataddr(&faddr->address, NULL, 0);
+				if (len > cw->foreign_addr)
+					cw->foreign_addr = len;
```
```suggestion
				cw->foreign_addr = MAX(cw->foreign_addr, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-			pos += xprintf("%d", s->fibnum);
-			offset += 7;
+			len = snprintf(NULL, 0, "%d", s->fibnum);
+			if (len > cw->fib)
+				cw->fib = len;
```
```suggestion
			cw->fib = MAX(cw->fib, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-					pos += xprintf("??");
-					offset += opt_w ? 46 : 22;
+					len = formataddr(&sp->laddr->address, NULL, 0);
+					if (len > cw->splice_address)
+						cw->splice_address = len;
```
```suggestion
					cw->splice_address = MAX(cw->splice_addres, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+				s->proto == IPPROTO_UDP) {
+				len = snprintf(NULL, 0,
+					"%" PRIu64, s->inp_gencnt);
+				if (len > cw->inp_gencnt)
+					cw->inp_gencnt = len;
```
```suggestion
				cw->inp_gencnt = MAX(cw->inp_gencnt, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+					s->state != TCPS_LISTEN))) {
+				len = snprintf(NULL, 0, "%u",
+					ntohs(faddr->encaps_port));
+				if (len > cw->encaps)
+					cw->encaps = len;
```
```suggestion
				cw->encaps = MAX(cw->encaps, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+				s->state != SCTP_LISTEN) {
+				len = snprintf(NULL, 0, "%s",
+					sctp_path_state(faddr->state));
+				if (len > cw->path_state)
+					cw->path_state = len;
```
```suggestion
				cw->path_state = MAX(cw->path_state, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
-						    sctp_conn_state(s->state));
+						len = snprintf(NULL, 0, "%s",
+							sctp_conn_state(s->state));
+						if (len > cw->conn_state)
+							cw->conn_state = len;
```
```suggestion
						cw->conn_state = MAX(cw->conn_state, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+							s->state < TCP_NSTATES) {
+							len = snprintf(NULL, 0, "%s",
+								tcpstates[s->state]);
+							if (len > cw->conn_state)
+								cw->conn_state = len;
```
```suggestion
							cw->conn_state = MAX(cw->conn_state, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+static void
+calculate_column_widths(struct col_widths *cw)
+{
+	cw->user		= 4;
+	cw->command		= 10;
+	cw->pid			= 3;
+	cw->fd			= 2;
+	cw->proto		= 5;
+	cw->local_addr	= 13;
+	cw->foreign_addr	= 15;
+	cw->pcb_kva		= 18;
+	cw->fib			= 3;
+	cw->splice_address	= 14;
+	cw->inp_gencnt	= 2;
+	cw->encaps		= 6;
+	cw->path_state	= 10;
+	cw->conn_state	= 10;
+	cw->stack		= TCP_FUNCTION_NAME_LEN_MAX;
+	cw->cc			= TCP_CA_NAME_MAX;
```
I suggest not trying to align the parts after the `=`.  But if you do, make sure you use spaces, or use tabstop=8.  As it is, they aren't actually aligned.

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+			len = snprintf(NULL, 0, "%lu", (u_long)xf->xf_uid);
+		else
+			len = snprintf(NULL, 0, "%s", pwd->pw_name);
+		if (len > cw->user)
+			cw->user = len;
```
```suggestion
		cw->user = MAX(cw->user, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		if (len > cw->user)
+			cw->user = len;
+		len = snprintf(NULL, 0, "%lu", (u_long)xf->xf_pid);
+		if (len > cw->pid)
+			cw->pid = len;
```
```suggestion
		cw->pid = MAX(cw->pid, len);
```

asomers commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+		if (len > cw->pid)
+			cw->pid = len;
+		len = snprintf(NULL, 0, "%d", xf->xf_fd);
+		if (len > cw->fd)
+			cw->fd = len;
```
```suggestion
		cw->fd = MAX(cw->fd, len);
```

asomers reviewed (changes requested):
Numeric columns like "FIB" "ENCAPS", and "FD" are now right-aligned, whereas they used to be left-aligned.  Is that intentional?

The "STACK" column is far wider than it needs to be.
That's probably because you use TCP_FUNCTION_NAME_LEN_MAX instead of calculating how much space is actually needed.  But that's overkill.  The constant is 32, whereas the longest stackname currently defined only uses 7 characters.  The same problem applies to the "CC" column.

The "opt_w" variable is now write-only.  You can simply remove it, but of course getopt() should still accept it.

daminrido139 commented:
Thanks a lot for the detailed review, @asomers !

1. The right-alignment of numeric columns was intentional for readability (similar to ps and netstat).
2. Great point on the STACK and CC column widths. I now compute them based on the actual observed lengths instead of using constant maximums.
3. Removed the opt_w variable while retaining it in the getopt() call.
4. Fixed the mid-line tab usage. my tabstop wasn’t set to 8 earlier; it is now corrected.
5. Reverted the usage() line to stay within the 80-character limit, as originally intended.
6. Applied all the suggested style changes: cuddled else, MAX() usage, line length limits, alignment fixes, etc.
7. Added a comment to clarify the use of the magic number 16.

As per your suggestion, I initially made the changes in a separate commit and pushed normally. However, I noticed a few unintended changes had crept in, so I had to amend and force-push to revert them. I apologize for that and will be more careful next time.

Please review the updated changes whenever you get a chance. Thanks again!

asomers reviewed (changes requested):
This is looking pretty good now!  I just have a few more comments regarding style.  Also, please ensure that every user-visible change is mentioned in the commit message.  That includes changing the alignment of numeric fields.

daminrido139 commented:
Hi @asomers ,
I’ve fixed those style issues and included the alignment change of the numeric fields in the commit message. Thank you for your valuable advice! Kindly review the changes when you have time and let me know the next steps.

asomers reviewed (changes requested):
Just one more small change: your commit message should include this line.  Other than that, it LGTM.

```
Pull Request:   https://github.com/freebsd/freebsd-src/pull/1720
```

daminrido139 commented:
@asomers done!!

asomers reviewed (approved):
Done.
At this point, if you were a committer, you should:

1. Squash your commits
2. Rebase on main
3. Force push to update the PR
4. push to main
5. If you get a conflict pushing to main, return to step 2.

However, since you aren't a committer, I will do those things for you.  And you can get back to work on libxo support.  Thanks for your contribution!

asomers reviewed (changes requested):
Testing revealed one more problem.  Your changes don't compile on i386 due to printf format specifier details.  It's a very common mistake.  It's surprising that they don't, because you used the exact same printf format specifier as the old code.  However, you switched from using vprintf to printf, which enables more compile-time checking.
```
/usr/home/somers/src/freebsd.org/src/usr.bin/sockstat/sockstat.c:1402:23: error: format specifies type 'unsigned long' but the argument has type 'kvaddr_t' (aka 'unsigned long long') [-Werror,-Wformat]
 1402 |                         printf(" 0x%16lx", s->pcb);
      |                                    ~~~~~   ^~~~~~
      |                                    %16llx
```
To reproduce, you can do `make TARGET_ARCH=i386 -j(nproc) buildworld` from the top of the git tree.

asomers reviewed (commented):
I'll approve it as-is, or with my suggestion.  Just let me know which you prefer.

Also, I just noticed that "-A" is undocumented.  We should probably fix that.  Not for this PR, though.

daminrido139 commented on usr.bin/sockstat/sockstat.c:
```diff
@@ -1399,7 +1400,8 @@ display_sock(struct sock *s, struct col_widths *cw, char *buf, size_t bufsize)
 		if (opt_A)
-			printf(" 0x%16lx", s->pcb);
+			printf(" 0x%*lx", cw->pcb_kva,
+				(u_long)(uintptr_t)s->pcb);
```
Yes, I noticed the use of the PRIu64 macro earlier. I went with your suggestion..  it's cleaner and more consistent now.

daminrido139 commented:
Hi @asomers ,
The "Allow edits by maintainers" checkbox is already ticked, so I’m not exactly sure what went wrong. I've sent you an invitation to be a collaborator on my fork, maybe that will help. Could you please check your GitHub inbox?
